### ISSUE TITLE
Hyperparameter “use_cmudict”

### ISSUE THREAD

**Issue by zuoxiang95** [keithito/tacotron#10]
Does anyone know the function of the hyperparameter  “use_cmudict”? Does it change the input words to phonemes?

**Comment by sonach** [comment]
Yes, it does.  It changes the words to phonemes randomly.

**Comment by zuoxiang95** [comment]
randomly? what do you means?

**Comment by keithito** [comment]
It will sample 50% of inputs and in those convert 50% of the words in those inputs to phonemes if the have unambiguous entries in CMUDict. See:

https://github.com/keithito/tacotron/blob/bc706654ad79765299c5b3077704cf138b64980d/datasets/datafeeder.py#L107
https://github.com/keithito/tacotron/blob/bc706654ad79765299c5b3077704cf138b64980d/datasets/datafeeder.py#L118

The intent was to be allow people to override pronunciations of specific words at synthesis time. There's a little more info in the README.

**Comment by zuoxiang95** [comment]
@keithito Thanks ~ Can i input phonemes to synthesis wav?

**Comment by keithito** [comment]
Yes, there's an example in the Miscellaneous section of the README.

**Comment by zuoxiang95** [comment]
Great! I will try it in this weekend~~

**Comment by keithito** [comment]
It looks like loss is still decreasing, just more slowly. If you increase the smoothing in tensorboard, it might be easier to see if this is the case. Does your eval output sound okay?

**Comment by zuoxiang95** [comment]
I think my output sound can be heard clear, but the pronunciation is still like a robot. I want it to be more natural.Can you give some advises  to modify the  model or the  hyperparameters?
here is my wav on 324000 steps.

[wav.zip](https://github.com/keithito/tacotron/files/1184990/wav.zip)

**Comment by sonach** [comment]
Set user_cmudict to True, and Let all the words turn to phonemes(which means not randomly) during training and synthesizing, will improve the audio quality.
For example, the input text for both training and synthesizing will turn to something like this: [{DH AH0} {B L UW1} {L AH0 G UW1 N} {IH1 Z} {AH0} {N AY1 N T IY1 N} {EY1 T IY0} {AH0 M EH1 R IH0 K AH0 N} {R OW0 M AE1 N S} {AE0 D V EH1 N CH ER0} {F IH1 L M} .] 
[tacotron_phoneme_step291000.zip](https://github.com/keithito/tacotron/files/1187206/tacotron_phoneme_step291000.zip)

**Comment by zuoxiang95** [comment]
@sonach  I am trying this. I set "user_cmudict = True",and "_p_cmudict =1". But there still has some words that we can not turn to phonemes. How did you deal with the situation?

**Comment by keithito** [comment]
Nice result, @sonach! 

@zuoxiang95: only words that have exactly one mapping in CMUDict get mapped to phonemes. You can allow words with more than one mapping by setting `keep_ambiguous=True` [here](https://github.com/keithito/tacotron/blob/bc706654ad79765299c5b3077704cf138b64980d/datasets/datafeeder.py#L59), but keep in mind that the phonemes that it uses will not always be the right ones. Sorry, this is not documented very well. When I have some time, I'll go through and document it better.

**Comment by sonach** [comment]
@keithito CMUDict functionality is very useful! 
Yes, the phonemes that it uses will not always be the right ones. 
And there is another problem, the words that don't appear in the CMUDict will remain unchanged(for example, the 'unpreceding' will be unchanged). 
For the second problem, we can use G2P(https://github.com/cmusphinx/g2p-seq2seq) to turn these words to phonemes.

**Comment by ElevenGameStudios** [comment]
@keithito First of all, thank's a lot for uploading your Tacotron version, the demo_server and cmudict option are really nice additions, compared to the other Tacotron verions already on github! I am already getting quite good results, even though my data set is pretty small (only 45 mins of speech).

@sonach Yes, I also considered using phonemes and the g2p-seq2seq model. Another user named Min (here https://gitter.im/voice-synthesis/tacotron) also stated that using phonemes leads to faster convergence of the model, which absolutely makes sense. regarding your two points:
(1) Why? If you are using keep_ambiguos=False, the used phonemes should always be the right ones, or am I missing something here?
(2) Yes. I am pretty sure that older tts systems must have used some algorithms to predict the correct pronounciation given an ambiguos word. One problem i've found is that the seq2seq model only runs on tensorflow 1.0, if they would convert it to the latest tensorflow version, it would be much easier to run it in realtime, e.g. as part of the demo_server. I once tried converting it myself, but my tensorflow knowledge/experience is limited and it didn't work out. If you or someone else is able to do that, I'd highly appreciate it.
Another interesting question would be whether it is useful to keep the stress information in the phoneme data? The original cmudict contains those stresses (e.g. "R EH1 D", "R IY1 D"), but the seq2seq model outputs phonemes without it ("R EH D", "R IY D"). One could either remove the numbers altogether or augment the seq2seq model by adding "default" stresses. Right now the easiest option seems to just drop them and use the seq2seq model to transcribe all training data words to create additional entries in the cmudict and then train regularly.

Anyway, it is nice to see that a lot of competent people from all over the world are experimenting with this machine learning algorithm! I am looking forward to all of our results. :D

**Comment by penelopeia** [comment]
@ElevenGameStudios you mentioned you are training on a very small data set, I am curious if you adjusted any of the hyper-parameters due to the size of your training data? Did you also have good results after 20k steps or so as reported? I am trying to also train on a small data set (~45 min), using the hyper-parameters set in the repo and unfortunately this keeps over fitting.

**Comment by ElevenGameStudios** [comment]
@penelopeia since i am training english voices, i never actually started from scratch, but used the pretrained model as initialization. therefore i had to keep the hyperparameters from the repo (i think generally speaking they are quite good in terms of model size vs. quality of synthesis).
while training it learns the features of the new voice pretty quickly and eventually starts overfitting. my most balanced voice results come from base model + ~5000 steps, so i can see how it is problematic to reach the mentioned 20k steps with just 45 min of voice. it seems there's a linear relationship between the amount of speech and the number of steps one can successfully train: if starting from scratch ~3 hours should be necessary... but it is always possible to create a base model from a bigger dataset of your given language and then let is adapt to features of different speakers in just a couple of steps.

**Comment by penelopeia** [comment]
@ElevenGameStudios thank you for the information! I will try what you have done, and train from the pretrained model. Hopefully I will get some interesting results. 

@keithito thanks for putting up your tacotron implementation, I was able to reproduce your results from the LJ speech dataset!

**Comment by ElevenGameStudios** [comment]
[ElevenSamples.zip](https://github.com/keithito/tacotron/files/1237345/ElevenSamples.zip)
Here are some samples of my results. i'm quite happy with voice1, but voice2's training set included certain microphone noise, so it's understandable that that stays in the output. voice3, the male speaker did not yet learn his attention properly, but i've heard that's happening all over the world. - seriously speaking, it probably needs a bigger dataset, that one too was about 45mins.

@penelopeia I can definitely recommend shaping the dataset like you want the final output to be. In my case I've had a lot of files of nearly the same length (close to the maximum length determined by max_iters) and splitting some of them (and the transcript) gave it a better capability to say single words as well as long sentences, etc.

**Comment by penelopeia** [comment]
[voice_samples.zip](https://github.com/keithito/tacotron/files/1243075/voice_samples.zip)
I created a new dataset that was 3-10 second clips. I also used the cmudict and was able to get some very interesting results!

@ElevenGameStudios thanks for the advice!

**Comment by keithito** [comment]
@penelopeia and @ElevenGameStudios, your samples sound really good!

**Comment by ElevenGameStudios** [comment]
@keithito Without the code you provided, they would probably not exist, so thank you again!

@penelopeia nice! I like your voice! I am currently working on a solution to synthesize from the trained models outside of a python environment. (C#/Unity3D which can even be compiled for web browsers. I already have a working setup for automatic lip synchronization for 3d models, that works reasonably well.) Maybe I can improve the griffin-lim part of the system in the process..

**Comment by 493238731** [comment]
Is there someone experiment Set use_cmudict=True when eval/decode stage? It maybe generate more natural wav.file?
Ｂｕｔ now it isn't support use_cmudict =True when eval/decode stage, it only inputs Letter to generage wav.file. How to change?

**Comment by zuoxiang95** [comment]
@493238731 You can input phonemes in eval stage. Only need train with setting "use_cmudict=True" in hparams.py and "_p_cmudict = 1" in datafeeder.py .

**Comment by 493238731** [comment]
@zuoxiang95 I know it can input phonemes in train stage. But it  doesn't call datafeeder.py in eval/demo_server stage.Exactly it calls synthersizer.py and synthesizer.py calls only text normalization(cleaner and text_to_sequence),not using cmudict.  So we should pre-process inputs to phonemes, you can see [line29](https://github.com/keithito/tacotron/blob/9fc433a8702a5e8392254f4c2cc6fe035c64e424/synthesizer.py#L29) to [line 30](https://github.com/keithito/tacotron/blob/9fc433a8702a5e8392254f4c2cc6fe035c64e424/synthesizer.py#L30)

**Comment by cozec** [comment]
Thank @keithito for the great work!

synthesized from Obama speech
[combined_obama-eval-354000.wav.zip](https://github.com/keithito/tacotron/files/1455979/combined_obama-eval-354000.wav.zip)

**Comment by keithito** [comment]
@cozec: that's a really good (and somewhat scary) result! What did you use for training data?

**Comment by cozec** [comment]
@keithito 
step1: ljspeech (default dataset): ep 206k
step2: obama:   ep 354k,
Obama data (1.6hour): downloaded from youtube; force alignment and segmentation.
Thank you again for you nice work! BTW, interested join our team? We have an opening for you!

**Comment by B1gM** [comment]
Hi there,

I'm trying to figure out how to use my own data to train off the pre-trained model, having a lot of trouble seeing how it works - any assistance would be much appreciated!

**Comment by cozec** [comment]
@B1gM Did you every try to stop training and start again from one particular checkpoint? If yes, then just replace the second step with your new data.

**Comment by mahdeto** [comment]
@cozec @keithito I wonder how do you do the fine tuning? do you change the learning rate? do you make certain layers untrainable? or do you just start with the pretrained model as your checkpoint and just train on your data set from that point?

**Comment by cozec** [comment]
this line is what i did:
"just start with the pretrained model as your checkpoint and just train on
your new data set from that point"

good luck.

**Comment by ZaraTahir** [comment]
How to start training of our model at a particular checkpoint?

**Comment by shartoo** [comment]
Hi there,thank you all for providing detail step of transfer learning  by T2. I tried to do this with Mandarin with audios from one female to another but result in blur result. How should a base model which will be transfered be trained like? For transfer learning task in image field  ,imaged with same label are many which equal to same word with many audio in TTS,should i prepare dataset like this,namely same words are spoken by multi-speaker ?
